lightly/cli/upload_cli.py

```diff
@@ -79,9 +79,11 @@ def _upload_cli(cfg, is_cli_call=True):
         with open(path_to_custom_metadata, 'r') as f:
             custom_metadata = json.load(f)
 
-    # determine the number of available cores
+    # set the number of workers if unset
     if cfg['loader']['num_workers'] < 0:
-        cfg['loader']['num_workers'] = cpu_count()
+        # set the number of workers to the number of CPUs available,
+        # but minimum of 8
+        cfg['loader']['num_workers'] = max(8, cpu_count())
```

Comment:
I think this line should not be in the `if` clause. If we move it outside and change it to: `cfg['loader']['num_workers'] = max(8, cfg['loader']['num_workers'])` the upload will always use at least 8 workers. If we don't do this then uploads might be slow if you manually set the number of workers to a low number (and I guess it breaks if you set it to 0 no?).

Reply:
I would do it differently:
- If the user sets `loader.num_workers`, it is taken and used. This is the idea of a config parameter. As a user I would want to have my parameters used just as I set them if I set them.
- Only if the user does not set the parameter, Lightly should recommend a sensible value, here `max(8, cpu_count())`.

Reply:
I agree that this is the better approach. Could you just verify that it works with 0 workers and I would still add an upper limit.

Reply:
> Could you just verify that it works with 0 workers

0 workers had always worked, it means not to use a separate thread, but use the same thread as the main program. I have also tested it, it works, just as before.

> I would still add an upper limit.

I don't see the sense of it as it
- Prevents the customer from specifying something they want conveniently.
- Does not prevent that customers overwrite it by changing the python code.
- Makes the upload for customers slower if they need it.

Reply:
> 0 workers had always worked, it means not to use a separate thread, but use the same thread as the main program. I have also tested it, it works, just as before.
 
Ok perfect!

> I don't see the sense of it as it

The idea of the upper limit is to avoid accidents that the user did not intend to happen. One of them is to overload the api by accident. The other one is to use many more cores on his machine than he intended to. If you use `cpu_count()` on a cluster you will run into trouble. This is also why the `ThreadPoolExecutor` class itself limits the numbers of worker threads if they are not specified:

https://github.com/python/cpython/blob/9b23f8f78fdb0d7eba016616ae7a97abbfc65aa6/Lib/concurrent/futures/thread.py#L134-L142

Reply:
@guarin You convinced me to limit it to 32, I implemented it.